pkg/controllers/garbagecollector/garbagecollector.go

```diff
@@ -0,0 +1,283 @@
+		return
+	}
+
+	go wait.Until(gb.worker, time.Second, stopCh)
+
+	<-stopCh
+}
+
+func (gb *GarbageCollector) addJob(obj interface{}) {
+	job := obj.(*v1alpha1.Job)
+	glog.V(4).Infof("Adding job %s/%s", job.Namespace, job.Name)
+
+	if job.DeletionTimestamp == nil && needsCleanup(job) {
+		gb.enqueue(job)
+	}
+}
+
+func (gb *GarbageCollector) updateJob(old, cur interface{}) {
+	job := cur.(*v1alpha1.Job)
+	glog.V(4).Infof("Updating job %s/%s", job.Namespace, job.Name)
+
+	if job.DeletionTimestamp == nil && needsCleanup(job) {
+		gb.enqueue(job)
+	}
+}
+
+func (gb *GarbageCollector) enqueue(job *v1alpha1.Job) {
+	glog.V(4).Infof("Add job %s/%s to cleanup", job.Namespace, job.Name)
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		glog.Errorf("couldn't get key for object %#v: %v", job, err)
+		return
+	}
+
+	gb.queue.Add(key)
+}
+
+func (gb *GarbageCollector) enqueueAfter(job *v1alpha1.Job, after time.Duration) {
+	key, err := controller.KeyFunc(job)
+	if err != nil {
+		glog.Errorf("couldn't get key for object %#v: %v", job, err)
+		return
+	}
+
+	gb.queue.AddAfter(key, after)
+}
+
+func (gb *GarbageCollector) worker() {
+	for gb.processNextWorkItem() {
+	}
+}
+
+func (gb *GarbageCollector) processNextWorkItem() bool {
+	key, quit := gb.queue.Get()
+	if quit {
+		return false
+	}
+	defer gb.queue.Done(key)
+
+	err := gb.processJob(key.(string))
+	gb.handleErr(err, key)
+
+	return true
+}
+
+func (gb *GarbageCollector) handleErr(err error, key interface{}) {
+	if err == nil {
+		gb.queue.Forget(key)
+		return
+	}
+
+	glog.Errorf("error cleaning up Job %v, will retry: %v", key, err)
+	gb.queue.AddRateLimited(key)
+}
+
+// processJob will check the Job's state and TTL and delete the Job when it
+// finishes and its TTL after finished has expired. If the Job hasn't finished or
+// its TTL hasn't expired, it will be added to the queue after the TTL is expected
+// to expire.
+// This function is not meant to be invoked concurrently with the same key.
+func (gb *GarbageCollector) processJob(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+
+	glog.V(4).Infof("Checking if Job %s/%s is ready for cleanup", namespace, name)
+	// Ignore the Jobs that are already deleted or being deleted, or the ones that don't need clean up.
+	job, err := gb.jobLister.Jobs(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if expired, err := gb.processTTL(job); err != nil {
+		return err
+	} else if !expired {
+		return nil
+	}
+
+	// The Job's TTL is assumed to have expired, but the Job TTL might be stale.
+	// Before deleting the Job, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	// The latest Job may have a different UID, but it's fine because the checks will be run again.
+	fresh, err := gb.vkClient.BatchV1alpha1().Jobs(namespace).Get(name, metav1.GetOptions{})
```

Comment:
In which case, we need to do a double check ?

Reply:
We can not guarantee the job TTL not changed between gc received the it and processing here.

Reply:
Mostly these codes are referred from k8s ttlafterfinished controller